src/Plugin/Payment/Method/PostfinancePaymentFormMethod.php

```diff
@@ -115,59 +70,34 @@ protected function doExecutePayment() {
       'CANCELURL' => $generator->generateFromRoute('payment_postfinance.response_cancel', array('payment' => $payment->id()), array('absolute' => TRUE)),
     );
 
-    // Generate SHA-IN signature
+    // Save the payment data if we are testing the module.
+    if (\Drupal::state()->get('postfinance.testing')) {
+      \Drupal::state()->set('postfinance.payment_data', $payment_data);
+    }
+
+    // Generate SHA-IN signature.
     $payment_data['SHASign'] = PostfinanceHelper::generateShaSign($payment_data, $this->pluginDefinition['sha_in_key']);
 
     // Generate payment link with correct query.
-    $payment_link = Url::fromUri($payment_config->get('payment_link'), array(
+    $payment_link = (Url::fromUri($payment_config->get('payment_link'), array(
+      'absolute' => TRUE,
       'query' => $payment_data,
-    ))->toString();
+    )));
 
-    // Redirect to generated payment link.
-    $response = new RedirectResponse($payment_link);
-    $listener = function (FilterResponseEvent $event) use ($response) {
-      $event->setResponse($response);
-      $event->stopPropagation();
-    };
-    $this->eventDispatcher->addListener(KernelEvents::RESPONSE, $listener, 999);
-
-    // Save payment
-    $payment->save();
+    return new PaymentExecutionResult(new Response($payment_link));
   }
 
   /**
-   * {@inheritdoc}
-   */
-  protected function getSupportedCurrencies() {
-    return TRUE;
-  }
-
-  /**
-   * {@inheritdoc}
-   */
-  protected function doCapturePaymentAccess(AccountInterface $account) {
-    // TODO: Implement doCapturePaymentAccess() method.
-  }
-
-  /**
-   * {@inheritdoc}
-   */
-  protected function doCapturePayment() {
-    // TODO: Implement doCapturePayment() method.
-  }
-
-  /**
-   * {@inheritdoc}
+   * Performs the actual payment execution.
    */
-  protected function doRefundPaymentAccess(AccountInterface $account) {
-    // TODO: Implement doRefundPaymentAccess() method.
+  protected function doExecutePayment() {
   }
 
   /**
    * {@inheritdoc}
    */
-  protected function doRefundPayment() {
-    // TODO: Implement doRefundPayment() method.
+  protected function getSupportedCurrencies() {
```

Comment:
We should check if postfinance has limitations here, and maybe an API or static list of allowed currencies. Can be a follow-up.